lib/puppet/provider/group/gpasswd.rb

```diff
@@ -0,0 +1,70 @@
+require 'puppet/provider/group/groupadd'
+
+Puppet::Type.type(:group).provide :gpasswd, :parent => Puppet::Type::Group::ProviderGroupadd do
+  require 'set'
+  require 'shellwords'
+
+  desc <<-EOM
+    Group management via `gpasswd`. This allows for local group
+    management when the users exist in a remote system.
+  EOM
+
+  commands  :addmember => 'gpasswd',
+            :delmember => 'gpasswd'
+
+  has_feature :manages_members unless %w{HP-UX Solaris}.include? Facter.value(:operatingsystem)
+
+  def addcmd
+    cmd = []
+    cmd << super.shelljoin
+
+    @resource[:members] and cmd << @resource[:members].map{ |x|
+      [ command(:addmember),'-a',x,@resource[:name] ].shelljoin
+    }
+
+    # NOTE: This will create the group but may not add some members to
+    # your group if they do not exist on the system. Any member that
+    # can be added will be added in an attempt to do "the right
+    # thing".
+    cmd.join(' && ')
+  end
+
+  def members
+    retval = @objectinfo.mem
+
+    if @resource[:members] and
+       @resource[:attribute_membership] == :minimum and
+       (@resource[:members] - @objectinfo.mem).empty?
+    then
+        retval = @resource[:members]
+    end
+
+    retval
+  end
+
+  def members=(members)
+    cmd = []
+    to_be_added = members.dup
+    if @resource[:attribute_membership] == :minimum
+      to_be_added = to_be_added.to_set.union(@objectinfo.mem)
+    else
+      to_be_removed = @objectinfo.mem - to_be_added
+      to_be_added = to_be_added - @objectinfo.mem
+
+      not to_be_removed.empty? and cmd << to_be_removed.map { |x|
```

Comment:
FWIW, the most readable form of this to my eyes would be `if !condition`. I also almost never use `unless` which slows down my reading except in the simplest of cases.

Also, when we went on a rubocop rampage a few months ago we *almost* merged a PR to enforce no use of `or`, `and`, and `not` (because of confusion between boolean and conditional operators). We still might do that someday, so for now best practice in the puppet code base is not to add more such uses.